src/main/kotlin/pl/edu/agh/gem/internal/service/ExpenseService.kt

```diff
@@ -0,0 +1,55 @@
+package pl.edu.agh.gem.internal.service
+
+import org.springframework.stereotype.Service
+import pl.edu.agh.gem.internal.client.CurrencyManagerClient
+import pl.edu.agh.gem.internal.client.GroupManagerClient
+import pl.edu.agh.gem.internal.model.currency.ExchangeRate
+import pl.edu.agh.gem.internal.model.expense.Expense
+import pl.edu.agh.gem.internal.persistence.ExpenseRepository
+import pl.edu.agh.gem.internal.validation.CostValidator
+import pl.edu.agh.gem.internal.validation.CurrenciesValidator
+import pl.edu.agh.gem.internal.validation.ExpenseDataWrapper
+import pl.edu.agh.gem.internal.validation.ParticipantValidator
+import pl.edu.agh.gem.model.GroupMembers
+import pl.edu.agh.gem.validator.ValidatorList.Companion.validatorsOf
+import pl.edu.agh.gem.validator.ValidatorsException
+import java.time.Instant
+
+@Service
+class ExpenseService(
+    private val groupManagerClient: GroupManagerClient,
+    private val currencyManagerClient: CurrencyManagerClient,
+    private val expenseRepository: ExpenseRepository,
+) {
+    private val expenseValidators = validatorsOf(
+        CostValidator(),
+        ParticipantValidator(),
+        CurrenciesValidator(),
+    )
+
+    fun getGroupMembers(groupId: String): GroupMembers {
+        return groupManagerClient.getGroupMembers(groupId)
+    }
+
+    fun create(groupMembers: GroupMembers, expense: Expense): Expense {
+        expenseValidators
+            .getFailedValidations(createExpenseDataWrapper(groupMembers, expense))
+            .takeIf { it.isNotEmpty() }
+            ?.also { throw ValidatorsException(it) }
+        val exchangeRate = getExchangeRate(expense.baseCurrency, expense.targetCurrency, expense.expenseDate)
```

Comment:
i dont sure if we should get exchange rate in that method, and not extract that to seperate function and paste that in creation of expense